### PR TITLE
Feature/add to up next after sync

### LINKED ISSUE #3211: Add new episodes to UpNext and trigger download after (not before) sync
### Description

Having two phones (devices probably too?) I have the first is setup to keep adding to UpNext (on top, and last, with 20 episodes limit), whilst I occasionally tick off the list items listening on the second device as well as listening to new episodes as these catch my attention (some are setup for UpNext on the 1st device).
On refresh (background disabled) I see update of podcast happening first, so episodes already listen to on 2nd device get added to UpNext (sometimes removing from the back of the list), manage to download, only to disappear a while later when sync happens.

I'd like to improve behaviour by adding to UpNext after the sync (which might actually bring some elements already)

### Screenshots or screen recording

_No response_

### Did you search for existing list?

- [x] I have searched for existing issues.

### PR DESCRIPTION
**Pull request by stic**
## Description
Attempt to defer addition to UpNext these episodes that are new (post update) but might have been already finished, archived or added to UpNext on another device. The update of UpNext got extracted as a separate function, filter was added (potentially causing some slowdown as it would check if episodes aren't finished, archived or already on the UpNext list).
Must admit I'm not 100% sure that UpNext sync will always finish by the time we return from sync(), so if anything logic there might be slightly broken (still don't got into duplicated or missing items on UpNext after some testing)

Fixes #3211

## Testing Instructions
1. To test, you would need two devices, one setup to add episodes to UpNext (another that doesn't)
2. Make sure you have background refresh / update turned off.
3. On the device that doesn't sync, pick an episode that should be added to UpNext (sometimes it takes a while for that to happen) on the device adding to UpNext
4. Listen to that episode(s)
5. Open and observe UpNext syncing on 1st device - before fix it would put the episode on UpNext, download and in a while remove
6. Postfix episode doesn't show up on UpNext, doesn't download, is seen as archived / played

## Screenshots or Screencast
Need to get my head around how to create these.

## Checklist
- [X] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [ ] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [X] I have considered whether it makes sense to add tests for my changes
- [ ] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [ ] Any jetpack compose components I added or changed are covered by compose previews
- [ ] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

#### I have tested any UI changes...
N/A

### CODE REVIEW

**Review by ashiagr — COMMENTED** [review]
I left two minor comments. 
It works! 🙂 We'll further monitor it in 7.78 beta.

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/refresh/RefreshPodcastsThread.kt [hunk 1]
```diff
@@ -291,16 +293,15 @@ class RefreshPodcastsThread(
         settings.setRefreshState(settings.getLastSuccessRefreshState() ?: RefreshState.Never)
     }
 
-    private fun updatePodcasts(result: RefreshResponse?): List<String> {
+    private data class AddedEpisodes(val episodeUuidsAdded: List<String>, val episodesToAddToUpNext: MutableList<Pair<AddToUpNext, PodcastEpisode>>)
```
[the remark this answers is not in the document]
Comment (stic) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/refresh/RefreshPodcastsThread.kt]: sure @ashiagr , refactored that only list is used..

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/refresh/RefreshPodcastsThread.kt [hunk 2]
```diff
@@ -347,12 +348,28 @@ class RefreshPodcastsThread(
...
+
         // Because we may not have refreshed for a while, we collect all the auto add to up next episodes
         // and run through them one by one sorted by their publish date. They are added to up next as if the action
         // was run right as they were published magically
         runBlocking {
             val upNextLimit = settings.autoAddUpNextLimit.value
             episodesToAddToUpNext.sortBy { it.second.publishedDate }
+            episodesToAddToUpNext.removeAll { runBlocking { true == episodeManager.findByUuid(it.second.uuid)?.isFinished ||
+                                                            true == episodeManager.findByUuid(it.second.uuid)?.isArchived ||
+                                                            playbackManager.upNextQueue.contains(it.second.uuid)} }
```
[the remark this answers is not in the document]
Comment (stic) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/refresh/RefreshPodcastsThread.kt]: sure, nice helper, also refactored a bit (i.e. moved logic of list of new episodes addition to UpNext to playbackManager as most it's there anyway), please let me know how you think it's looking now

**Comment by ashiagr** [comment]
@stic can you run 

`./gradlew spotlessApply` for linter to pass?

**Comment by stic** [comment]
Hi @ashiagr my refactoring changes got reverted on the way to PR, let me fix it later today (I think I've pressed sync fork whilst on branch in GitHub after I've pushed changes)..

**Comment by stic** [comment]
All merged in as it should be, please have a look

**Review by ashiagr — APPROVED** [review]

**Pull request merged**